solr/core/src/java/org/apache/solr/handler/component/CombinedQueryComponent.java

```diff
@@ -0,0 +1,590 @@
+          && Boolean.TRUE.equals(segmentTerminatedEarly)) {
+        rb.rsp
+            .getResponseHeader()
+            .remove(SolrQueryResponse.RESPONSE_HEADER_SEGMENT_TERMINATED_EARLY_KEY);
+        rb.rsp
+            .getResponseHeader()
+            .add(
+                SolrQueryResponse.RESPONSE_HEADER_SEGMENT_TERMINATED_EARLY_KEY,
+                segmentTerminatedEarly);
+      }
+    }
+    if (maxHitsTerminatedEarly) {
+      rb.rsp
+          .getResponseHeader()
+          .add(SolrQueryResponse.RESPONSE_HEADER_MAX_HITS_TERMINATED_EARLY_KEY, Boolean.TRUE);
+      if (approximateTotalHits > 0) {
+        rb.rsp
+            .getResponseHeader()
+            .add(
+                SolrQueryResponse.RESPONSE_HEADER_APPROXIMATE_TOTAL_HITS_KEY, approximateTotalHits);
+      }
+    }
+  }
+
+  /**
+   * Combines and sorts documents from multiple shards to create the final result set. This method
+   * uses a combiner strategy to merge shard responses, then sorts the resulting documents using a
+   * priority queue based on the request's sort specification. It handles pagination (offset and
+   * count) and calculates the maximum score for the response.
+   *
+   * @param rb The ResponseBuilder containing the request and context, such as sort specifications.
+   * @param shardDocMap A map from shard addresses to the list of documents returned by each shard.
+   * @param responseDocs The final response document list, which will be populated with null
+   *     placeholders and have its max score set.
+   * @return A map from document IDs to the corresponding ShardDoc objects for the documents in the
+   *     final sorted page of results.
+   */
+  protected Map<Object, ShardDoc> createShardResult(
+      ResponseBuilder rb, Map<String, List<ShardDoc>> shardDocMap, SolrDocumentList responseDocs) {
+    String algorithm =
+        rb.req.getParams().get(CombinerParams.COMBINER_ALGORITHM, CombinerParams.DEFAULT_COMBINER);
+    QueryAndResponseCombiner combinerStrategy =
+        QueryAndResponseCombiner.getImplementation(algorithm, combiners);
+    List<ShardDoc> combinedShardDocs = combinerStrategy.combine(shardDocMap, rb.req.getParams());
+    Map<String, ShardDoc> shardDocIdMap = new HashMap<>();
+    shardDocMap.forEach(
+        (shardKey, shardDocs) ->
+            shardDocs.forEach(shardDoc -> shardDocIdMap.put(shardDoc.id.toString(), shardDoc)));
+    Map<Object, ShardDoc> resultIds = new HashMap<>();
+    float maxScore = 0.0f;
+    Sort sort = rb.getSortSpec().getSort();
+    SortField[] sortFields;
+    if (sort != null) {
+      sortFields = sort.getSort();
+    } else {
+      sortFields = new SortField[] {SortField.FIELD_SCORE};
+    }
+    final ShardFieldSortedHitQueue queue =
+        new ShardFieldSortedHitQueue(
+            sortFields,
+            rb.getSortSpec().getOffset() + rb.getSortSpec().getCount(),
+            rb.req.getSearcher());
+    combinedShardDocs.forEach(queue::insertWithOverflow);
```

Comment:
Could you help me understand why the combined shard docs need to go through the sorting queue here?